lib/phenotype/versioned_app.rb

```diff
@@ -0,0 +1,48 @@
+module Phenotype
+  class VersionedApp
+    attr_accessor :strategies, :routes
+    attr_writer :no_strategy_response, :too_many_strategies_response, :version_not_found_response
+
+    def initialize(strategies: [])
+      @routes = {}
+      @strategies = strategies
+    end
+
+    def add_strategy(strat)
+      strategies.push(strat)
+    end
+
+    def add_version(version:, &block)
+      routes[version] = block
+    end
+
+    def call(env)
+      strats = strategies.find_all { |strat| strat.version(env) }
+      return no_strategy_response.call(env) if strategies.empty?
```

Comment:
So if we have a 406 status code, the default middleware would execute the default route? That should work. We could also give it a silly, somewhat shameful name like http://ruby-doc.org/stdlib-1.9.3/libdoc/minitest/unit/rdoc/MiniTest/Unit/TestCase.html#method-c-i_suck_and_my_tests_are_order_dependent-21